crates/services/txpool_v2/src/pool.rs

```diff
@@ -224,30 +227,65 @@ where
-        let extracted_transactions = self
-            .selection_algorithm
-            .gather_best_txs(
-                Constraints {
-                    max_gas: self.config.max_block_gas,
-                },
-                &mut self.storage,
-            )?
+        constraints: Constraints,
+    ) -> Vec<ArcPoolTx> {
+        self.selection_algorithm
+            .gather_best_txs(constraints, &mut self.storage)
             .into_iter()
             .map(|storage_entry| {
                 self.update_components_and_caches_on_removal(iter::once(&storage_entry));
 
                 storage_entry.transaction
             })
-            .collect::<Vec<_>>();
-
-        Ok(extracted_transactions)
+            .collect::<Vec<_>>()
     }
 
-    #[allow(dead_code)]
-    pub fn find_one(&self, tx_id: &TxId) -> Option<&PoolTransaction> {
+    pub fn find_one(&self, tx_id: &TxId) -> Option<ArcPoolTx> {
         Storage::get(&self.storage, self.tx_id_to_storage_id.get(tx_id)?)
-            .map(|data| &data.transaction)
+            .map(|data| data.transaction.clone())
+    }
+
+    pub fn contains(&self, tx_id: &TxId) -> bool {
+        self.tx_id_to_storage_id.contains_key(tx_id)
+    }
+
+    pub fn iter_tx_ids(&self) -> impl Iterator<Item = &TxId> {
+        self.tx_id_to_storage_id.keys()
+    }
+
+    /// Remove transaction but keep its dependents.
+    /// The dependents become executables.
+    pub fn remove_transaction(&mut self, tx_ids: Vec<TxId>) {
+        for tx_id in tx_ids {
+            if let Some(storage_id) = self.tx_id_to_storage_id.remove(&tx_id) {
+                let dependents: Vec<S::StorageIndex> =
+                    self.storage.get_direct_dependents(storage_id).collect();
+                let Some(transaction) = self.storage.remove_transaction(storage_id)
+                else {
+                    debug_assert!(false, "Storage data not found for the transaction");
+                    tracing::warn!(
+                        "Storage data not found for the transaction during `remove_transaction`."
+                    );
+                    continue
+                };
+                for dependent in dependents {
+                    let Some(storage_data) = self.storage.get(&dependent) else {
+                        debug_assert!(
+                            false,
+                            "Dependent storage data not found for the transaction"
+                        );
+                        tracing::warn!(
+                            "Dependent storage data not found for \
+                            the transaction during `remove_transaction`."
+                        );
+                        continue
+                    };
+                    self.selection_algorithm
+                        .new_executable_transaction(dependent, storage_data);
+                }
+                self.update_components_and_caches_on_removal(iter::once(&transaction));
```

Comment:
Just noticed that all tests in `crates/services/txpool_v2/src/tests/mod.rs` are still green if I remove these lines. Seems like tests don't ensure that we promote dependents to executables.

Maybe there are tests for that in different place.